### PR TITLE
[BEHAVIORAL] Wire HTTP and gRPC Check endpoints to GraphResolver

### LINKED ISSUE #173: [Enhancement] Wire GraphResolver to remaining API endpoints (gRPC, Expand, ListObjects)
## Summary

The HTTP Check API has been successfully wired to use the fully-implemented GraphResolver (PR #172 merged the CockroachDB fix, and subsequent work wired the resolver). This improved PostgreSQL compatibility tests from **43 passed / 207 failed** to **268 passed / 232 failed** - a gain of 225 more tests passing.

## Completed Work

- ✅ HTTP `POST /stores/{store_id}/check` now uses `GraphResolver::check()` for full graph resolution
- ✅ Added `From<DomainError> for ApiError` error conversion
- ✅ Support for contextual tuples in check requests
- ✅ Support for CEL context in check requests
- ✅ Store existence validation before check
- ✅ Updated all unit/integration/security tests with proper authorization model setup

## Remaining Work

The following endpoints need similar wiring to use the domain layer:

### Priority 1: gRPC Check API
- [ ] Wire `OpenFgaGrpcService::check()` to `GraphResolver::check()`
- [ ] Currently only batch_check is implemented in gRPC

### Priority 2: Expand API
- [ ] Implement actual tree expansion logic
- [ ] Current stub returns `{ tree: None }`
- [ ] Located at `routes.rs:expand()` (line ~815)

### Priority 3: ListObjects API  
- [ ] Implement object enumeration logic
- [ ] Current stub returns `{ objects: [] }`
- [ ] Located at `routes.rs:list_objects()` (line ~1168)

### Priority 4: Edge Cases
- [ ] Tuple read with userset references
- [ ] Tuple write validation with type constraints
- [ ] CEL condition evaluation in checks

## Test Results After Fix

```
PostgreSQL: 268 passed, 232 failed
```

Passing test sections:
- Section 2: Health/Info API (9/9)
- Section 3: Store CRUD (8/8)
- Section 6: Authorization Model Operations (6/6)
- Section 18: Rate Limiting (3/3)
- Section 34: Streamed ListObjects (5/5)
- Section 35: Modular Models (6/7)
- Sections 4-5: Model Validation (partial)

## Technical Notes

The `GraphResolver` in `rsfga-domain` is fully implemented with:
- All 6 userset types (direct, computed, tuple-to-userset, union, intersection, exclusion)
- Parallel branch execution with short-circuiting
- Cycle detection and depth limiting (max 25)
- CEL condition evaluation with expression caching
- Check result caching (optional)
- 4816 lines of tests

The main work is wiring this existing implementation to the API layer handlers.

## Related Files

- `crates/rsfga-api/src/http/routes.rs` - HTTP handlers
- `crates/rsfga-api/src/grpc/service.rs` - gRPC handlers
- `crates/rsfga-domain/src/resolver/graph_resolver.rs` - The resolver
- `crates/rsfga-server/src/handlers/` - Server layer handlers

### PR DESCRIPTION
**Pull request by julianshen**
## Summary

- Wires HTTP Check endpoint to use `GraphResolver.check()` for full graph traversal instead of direct tuple lookup
- Wires gRPC Check endpoint to use `GraphResolver.check()` with proper contextual tuple handling
- Adds `From<DomainError>` implementation for API error conversion
- Adds `resolver` field to `OpenFgaGrpcService` struct
- Updates all tests to include authorization model setup (required by GraphResolver for type/relation lookups)

Fixes #173

## Test plan

- [x] All existing tests pass (166 tests)
- [x] HTTP Check now performs full graph traversal
- [x] gRPC Check now performs full graph traversal
- [x] Security tests verify injection protection still works
- [x] Integration tests verify end-to-end flows work correctly

## Note

The Expand and ListObjects APIs remain as stubs - they require new domain layer methods (`GraphResolver.expand()` and `GraphResolver.list_objects()`) which are beyond the scope of this PR. A follow-up issue has been created to track this work.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Centralized domain-to-protocol error classification exposed for API responses.
  * Resolver-driven authorization check flow integrated into HTTP and gRPC endpoints.

* **Bug Fixes**
  * Improved mapping of domain errors to client-friendly error responses.
  * Removed duplicate/incomplete validation by delegating checks to the resolver, improving consistency.

* **Tests**
  * Test setups now pre-populate authorization models to ensure reliable check/write scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: crates/rsfga-api/src/grpc/service.rs [hunk 1]
```diff
@@ -392,42 +446,53 @@ impl<S: DataStore> OpenFgaService for OpenFgaGrpcService<S> {
...
+                                    .context
+                                    .and_then(|s| prost_struct_to_hashmap(s).ok()),
+                            )
+                        } else {
+                            ContextualTuple::new(&tk.user, &tk.relation, &tk.object)
+                        }
+                    })
+                    .collect()
+            })
+            .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (julianshen) [reply crates/rsfga-api/src/grpc/service.rs]: Fixed in commit a9429c6.

Changed the contextual tuple conversion to properly propagate errors instead of silently ignoring them with `.ok()`. The implementation now:

1. Uses `.transpose()` to convert `Option<Result<T, E>>` to `Result<Option<T>, E>`
2. Maps errors to `Status::invalid_argument` with a descriptive message including the tuple identifier
3. Collects results using `collect::<Result<Vec<_>, Status>>()` to fail fast on first error

This prevents potential authorization bypass when condition contexts contain invalid values (NaN, Infinity).

— Fixed by Claude

File: crates/rsfga-api/src/http/routes.rs [hunk 2]
```diff
@@ -190,6 +192,55 @@ impl From<StorageError> for ApiError {
...
+                    ApiError::internal_error("internal error during authorization check")
+                }
+            }
+            _ => {
+                error!("Domain error: {}", err);
+                ApiError::internal_error("internal error during authorization check")
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (julianshen) [reply crates/rsfga-api/src/http/routes.rs]: Fixed in commit a9429c6.

Created a shared `crate::errors` module with:

1. **`DomainErrorKind` enum** - Protocol-agnostic error classification:
   - `InvalidInput(String)` → 400/INVALID_ARGUMENT
   - `NotFound(String)` → 404/NOT_FOUND  
   - `Timeout(String)` → 504/DEADLINE_EXCEEDED
   - `Internal(String)` → 500/INTERNAL

2. **`classify_domain_error()` function** - Centralizes the error classification logic

3. **Unit tests** - 7 tests covering all error types

Both HTTP (`From<DomainError> for ApiError`) and gRPC (`domain_error_to_status`) now use this shared utility, ensuring consistent error handling and easier maintenance.

— Fixed by Claude
